src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,87 +87,102 @@
    pool :- jruby-schemas/pool-queue-type]
   (= expected-pool-size (count (.getRegisteredElements pool))))
 
-(schema/defn ^:always-validate
-  swap-and-drain-pool!
-  "Replace the current pool with a new pool and drain the old pool,
-  optionally refilling the new pool with fresh jrubies."
+(schema/defn collect-all-jrubies :- [JRubyInstance]
+  "Locks the pool and borrows all the instances"
+  [pool-state :- jruby-schemas/PoolState]
+  (let [pool-size (:size pool-state)
+        pool (:pool pool-state)
+        borrow-fn (fn [_] (jruby-internal/borrow-from-pool!*
+                           jruby-internal/borrow-without-timeout-fn
+                           pool))]
+    (.lock pool)
+    (try
+      (mapv borrow-fn (range pool-size))
+      (catch Exception e
+        (.clear pool)
+        (.insertPill pool (PoisonPill. e))
+        (throw (IllegalStateException.
+                "There was a problem borrowing a JRubyInstance from the pool."
+                e)))
+      (finally
+        (.unlock pool)))))
+
+(schema/defn cleanup-and-refill-pool
+  "Cleans up the given instances and optionally refills the pool with
+  new instances. Should only be called from the modify-instance-agent"
   [pool-context :- jruby-schemas/PoolContext
-   old-pool-state :- jruby-schemas/PoolState
-   new-pool-state :- jruby-schemas/PoolState
+   pool-state :- jruby-schemas/PoolState
+   old-instances :- [JRubyInstance]
    refill? :- schema/Bool]
-  (let [{:keys [config]} pool-context
-        pool-state-atom (jruby-internal/get-pool-state-container pool-context)
-        new-pool (:pool new-pool-state)
-        old-pool (:pool old-pool-state)
-        old-pool-size (:size old-pool-state)
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)
+        new-instance-ids (map inc (range pool-size))
```

Comment:
Hmm, good point.  I *think* the ids were only used just for troubleshooting / diagnostics purposes.  Like you could see "Flushing instance [id]" in the log when an instance was flushed.  So I don't know that the current set of changes on this PR would functionally break anything other than the usability of the log messages, which would still be a shame.

I remember debating with Chris at one point about whether we should just have an atomic id counter in the pool that we increment by one and hand off to the instance each time we create a new one.  Maybe this would be a good time to do that now?  I'd be okay with doing that as a follow-on ticket to this one, though, to avoid holding this one up further - but maybe one we should do before we do the next jruby-utils release?

@rlinehan, @jpinsonault - WDYT?